app/models/token/hashed_token.rb

```diff
@@ -0,0 +1,71 @@
+# Redmine - project management software
+# Copyright (C) 2006-2009  Jean-Philippe Lang
+# Adapted to fit needs for mOTP
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+require_dependency 'token/base'
+
+module Token
+  class HashedToken < Base
+    # Allow access to the plain value during initial access / creation of the token
+    attr_reader :plain_value
+
+    class << self
+      def create_and_return_value(user)
+        create(user: user).plain_value
+      end
+
+      ##
+      # Find a token from the token value
+      def find_by_plaintext_value(input)
+        find_by(value: hash_function(input))
+      end
+
+      private
+
+      ##
+      # Use a fixed salt for hashing token values.
+      # We still want to be able to index the hash value for fast lookups,
+      # so we need to determine the hash without knowing the associated user (and thus its salt) first.
+      def token_salt
+        '28f939460f6f852268a534f449928af54026af6c16aebf04f5975307b9d72de389f0'.freeze
```

Comment:
Does it make sense to hard-code this salt in here? Shouldn't we perhaps at least use a different salt per instance? Maybe based on the SECRET_KEY_BASE or something?